pallets/thea/src/validation.rs

```diff
@@ -237,16 +256,51 @@ fn get_finalized_head(url: &str) -> Result<String, &'static str> {
 	"params": []
 	});
 	let mut result =
-		send_request("get_finalized_head", url, body.to_string().as_str())?.to_string();
+		send_request("get_finalized_head", destination, body.to_string().as_str())?.to_string();
 	result = result.replace('\"', "");
 	log::debug!(target:"thea","Finalized head: {:?}",result);
 	Ok(result)
 }
 
+pub fn resolve_destination_url(destination: Destination, counter: i32) -> String {
+	if destination == Destination::Aggregator {
+		return AGGREGRATOR_URL.to_string()
+	}
+	let url = match (destination, counter) {
+		(Destination::Solochain, 0) => "http://localhost:9944",
+		(Destination::Solochain, 1) => MAINNET_URL,
+		(Destination::Parachain, 0) => "http://localhost:8844",
+		(Destination::Parachain, 1) => PARACHAIN_URL,
+		_ => AGGREGRATOR_URL,
+	};
+	log::debug!(target:"thea","Resolving {:?}: {:?} to {:?}",destination,counter,url);
+	url.to_string()
+}
+
 pub fn send_request(
 	log_target: &str,
-	url: &str,
+	destination: Destination,
 	body: &str,
+) -> Result<serde_json::Value, &'static str> {
+	for try_counter in 0..2 {
+		match create_and_send_request(
+			log_target,
+			body,
+			&resolve_destination_url(destination, try_counter),
+		) {
+			Ok(value) => return Ok(value),
+			Err(err) => {
+				log::error!(target:"thea","Error querying {:?}: {:?}",log_target, err);
+			},
+		}
+	}
```

Comment:
If I understood this logic correctly - if request to localhost is unsuccessful because of multiple possible reasons (unable to serialize body length on 308 line or on request 317 line or extract/decode response etc present in `create_and_send_request` method) we are  doing aka backup request to `MAINNET_URL`? It looks like attempt to guess endpoint (e.g. if locallhost is unavailable than try `MAINNET_URL`), to be able figure out environment dynamically (basically to be ready for any environment) but because of multiple errors which could happen in `create_and_send_request` method, or whatever reason because of which loclhost is unavailable - it could bring unpredictable behavior (e.g. accessing mainnet data from dev/testing environment when for some reason "http://localhost:9944" address is unavailable). 
Does it make sense?

IMO url should be passed by CLI args, basically configured at the beginning of app run.

Reply:
it can also be injected via RPC if it's runtime only

Reply:
Issue #859 created to cover this separately. Resolving this comment for now.